File: dart/network/lib/handshake.dart
import 'dart:math';
import 'dart:typed_data';

import 'package:blockchain_crypto/ed25519.dart';
import 'package:blockchain_crypto/utils.dart';
import 'package:blockchain_network/util.dart';

class HandshakeResult {
  final PeerId peerId;

  HandshakeResult(this.peerId);
}

Future<HandshakeResult> handshake(BytesReader reader, BytesWriter writer,
    Ed25519KeyPair localPeerKey, Uint8List magicBytes) async {
  await writer(magicBytes);
  final remoteMagicBytes = await reader(32);
  if (!magicBytes.sameElements(Uint8List.fromList(remoteMagicBytes)))
    throw new ArgumentError("Invalid remote magic bytes");
  await writer(localPeerKey.vk);
  final remoteVk = await reader(32);
  final localChallenge = List.generate(32, (i) => Random().nextInt(256));
  await writer(localChallenge);
  final remoteChallenge = await reader(32);
  final localSignature = await ed25519.sign(remoteChallenge, localPeerKey.sk);
  await writer(localSignature);
  final remoteSignature = await reader(32);
  if (!(await ed25519.verify(remoteSignature, localChallenge, remoteVk)))
    throw ArgumentError("Invalid remote signature");
  final peerId = Uint8List.fromList(remoteVk);
  return HandshakeResult(peerId);
}

File: scala/node/src/main/scala/com/giraffechain/p2p/Handshake.scala
package com.giraffechain.p2p

import com.giraffechain.crypto.CryptoResources
import cats.implicits.*
import com.giraffechain.models.PeerId
import com.giraffechain.utility.*
import cats.MonadThrow
import cats.data.OptionT
import cats.effect.Sync
import cats.effect.std.Random
import fs2.Chunk
import fs2.io.net.Socket
import scodec.bits.ByteVector

object Handshake:
  def run[F[_]: Sync: Random: CryptoResources](
      socket: Socket[F],
      magicBytes: Array[Byte],
      sk: Array[Byte],
      vk: Array[Byte]
  ): F[PeerId] =
    run(
      length =>
        OptionT(socket.read(length)).getOrRaise(new IllegalArgumentException(s"Expected $length bytes")).map(_.toArray),
      data => socket.write(Chunk.array(data)),
      magicBytes,
      sk,
      vk
    )

  def run[F[_]: Sync: Random: CryptoResources](
      read: Int => F[Array[Byte]],
      write: Array[Byte] => F[Unit],
      magicBytes: Array[Byte],
      sk: Array[Byte],
      vk: Array[Byte]
  ): F[PeerId] =
    for {
      ed25519 <- CryptoResources[F].ed25519.pure[F]
      exchanger = exchange(read, write)
      _ <- exchanger(magicBytes).ensure(new IllegalArgumentException("MagicBytes Mismatch"))(
        java.util.Arrays.equals(_, magicBytes)
      )
      remoteVk <- exchanger(vk)
      localChallenge <- Random[F].nextBytes(32)
      remoteChallenge <- exchanger(localChallenge)
      localSignature <- ed25519.useSync(_.sign(sk, remoteChallenge))
      remoteSignature <- exchanger(localSignature)
      signatureIsValid <- ed25519.useSync(_.verify(remoteSignature, localChallenge, remoteVk))
      _ <- Sync[F].raiseWhen(!signatureIsValid)(new IllegalArgumentException("Invalid Signature"))
    } yield PeerId(ByteVector(remoteVk).toBase58)

  private def exchange[F[_]: MonadThrow](read: Int => F[Array[Byte]], write: Array[Byte] => F[Unit])(
      data: Array[Byte]
  ): F[Array[Byte]] =
    write(data) >> read(data.length)
